std/range/package.d

```diff
@@ -4125,32 +4125,60 @@ pure unittest
     Generate lockstep's opApply function as a mixin string.
     If withIndex is true prepend a size_t index to the delegate.
 */
-private string lockstepMixin(Ranges...)(bool withIndex)
+private string lockstepMixin(Ranges...)(bool withIndex, bool reverse)
 {
     import std.format : format;
 
     string[] params;
     string[] emptyChecks;
     string[] dgArgs;
     string[] popFronts;
+    string indexDef;
+    string indexInc;
 
     if (withIndex)
     {
         params ~= "size_t";
         dgArgs ~= "index";
+        if (reverse)
+        {
+            indexDef = q{
+                size_t index = ranges[0].length-1;
+                enforce(_stoppingPolicy == StoppingPolicy.requireSameLength, "lockstep can only be used with foreach_reverse when stoppingPolicy == requireSameLength");
```

Comment:
> Test please ;-)

A test has been added. 

>>break symmetry with regular foreach_reverse (indexes would be wrong)
>
> ```
> foreach_reverse (i, e1, e2; lockstep([1,2,3], [4,5])) {
>   // (2,3,5), (1, 4, 2)
> }
> ```

Yup, that breaks indexing - 5 is not element 2 in the array [4,5]. Writing it slightly differently, both asserts below should hold:

```
auto a = [1,2,3];
auto b = [4,5];
foreach_reverse(i, e1, e2; lockstep(a,b)) {
    assert(e1 == a[i]);
    assert(e2 == b[i]);
}
```

> I think I will just stop annoying you

Please, you're not annoying - I am very glad you sparked the discussion, as it made me consider alternative solutions. All solutions have trade-offs, and I feel the ones I've chosen best fit the D philosophy, but that is certainly debatable. 

I've given some of the reasons I've chosen what I have, above. Another reason is that anyone can easily add indexing with iota or an explicit counter variable - this way it's an opt-in cost, and you have to consider the trade-offs yourself.

Lastly, there is no "intuitive" way for it to behave, so the default behavior would confuse some portion of users no matter what. Throwing an exception is not optimal, but at least it gives an understandable error message instead of iterating over a different set of values than the user intended.